Fixes #27728 - Add new bash completion

ofedoren opened the pull request.

ofedoren commented:
@mbacovsky, updated. I've noticed that we don't load settings from hammer-cli/config/cli_config.template.yml file. See my addition: https://github.com/theforeman/hammer-cli/blob/f57a352ad674da263f3c2434878843abbd7fc8bd/lib/hammer_cli/settings.rb#L20

@shiramax, If you run ./bin/hammer-complete you will see that the hammer will be initialized fully only if `~/.cache/hammer_completion.yml` file is missing. Also, on dev env if you want to speed-up autocompletion, add $path_to_hammer_bin on the first place in $PATH environment variable.

ofedoren commented:
@shiramax, @mbacovsky, I think it's ready to be merged.

Also, the completion time is cca 0.17s.

mbacovsky commented on bin/hammer-complete:
```diff
@@ -0,0 +1,55 @@
+  cache_file
+end
+
+completion_cache_file = cache_file || '~/.cache/hammer_completion.yml'
```
@ofedoren it may spare some cycles if we make it more opinionated and define it as `ENV['HAMMER_COMPLETION_CACHE'] || '~/.cache/hammer_completion.yml'`. No idea how much time we spend with the config loading and parsing. We could do similarly for the hammer command name (`HAMMER`).

mbacovsky commented on bin/hammer-complete:
```diff
@@ -0,0 +1,55 @@
+dict = HammerCLI::Bash::Completion.load_description(completion_cache_file)
+
+# regenerate and reload cache if it expired
+sha1sum = `sha1sum #{dict['expire']['file']}` if File.exist?(dict['expire']['file'])
```
When we expire the cache based on apipie cache couldn't we remove the completion cache with apipie cache update? That way we could avoid re-computing sha of the long file with every <tab>. Would that be possible @ofedoren?

ofedoren commented:
@mbacovsky, updated. I've removed sha computing, so the new completion cache is created when reloading apipie cache or if the completion cache file doesn't exist. Also removed using of `which` command.

xprazak2 commented on bin/hammer-complete:
```diff
@@ -0,0 +1,28 @@
+
+require 'yaml'
+
+completion_cache_file = ENV['HAMMER_COMPLETION_CACHE'] || '~/.cache/hammer_completion.yml'
```
`cli_config.yml` is loaded into ENV and we pull the value for cache location from there or use default, right?

xprazak2 commented:
Works nicely for me, just one thing - when I was randomly testing commands, I got a strange value completed for the os command, just for the os-params-attrs:

![os-params-attrs](https://user-images.githubusercontent.com/2664090/70790773-07b7b000-1d96-11ea-8512-0b2dba814f8c.png)

xprazak2 commented:
Also test failures seem related.

ofedoren commented:
@xprazak2, updated. The strange value is fixed according to our discussion and the tests are fixed :)

xprazak2 commented:
Thanks everyone!

tbrisker commented:
Build is now failing: https://community.theforeman.org/t/hammer-cli-master-package-release-29-failed/16777

xprazak2 commented:
Packaging fix: https://github.com/theforeman/foreman-packaging/pull/4539
